sdk/storage/azure-storage-blob/samples/client_side_encryption_keyvault.py

```diff
@@ -0,0 +1,123 @@
+
+from azure.keyvault.keys.crypto import CryptographyClient, KeyWrapAlgorithm
+from azure.keyvault.keys import KeyVaultKey
+from azure.keyvault.secrets import SecretClient
+
+from azure.storage.blob import BlobServiceClient, BlobType
+
+# Environment variable keys whic must be set to run this sample
+STORAGE_URL = 'AZURE_STORAGE_URL'
+KEYVAULT_URL = 'AZURE_KEYVAULT_URL'
+CLIENT_ID = 'AZURE_CLIENT_ID'
+CLIENT_SECRET = 'AZURE_CLIENT_SECRET'
+TENANT_ID = 'AZURE_TENANT_ID'
+
+def get_env_var(key):
+    try:
+        return os.environ[key]
+    except KeyError:
+        print('{} must be set.'.format(key))
+        sys.exit(1)
+
+def make_resource_name(prefix):
+    return '{}{}'.format(prefix, str(uuid.uuid4()).replace('-', ''))
+
+class KeyWrapper:
+    """ Class that fulfills the interface used by the storage SDK's
+        automatic client-side encyrption and decryption routines. """
+
+    def __init__(self, kek, credential):
+        self.kek = kek
+        self.kid = kek._properties._id
+        self.algorithm = KeyWrapAlgorithm.aes_256
+
+    def wrap_key(self, key):
+        crypto_client = CryptographyClient(self.kek, credential)
+        wrapped = crypto_client.wrap_key(key=key, algorithm=self.algorithm)
+        return wrapped.encrypted_key
+
+    def unwrap_key(self, key, algorithm):
+        crypto_client = CryptographyClient(self.kek, credential)
+        unwrapped = crypto_client.unwrap_key(encrypted_key=key, algorithm=algorithm)
+        return unwrapped.key
+
+    def get_key_wrap_algorithm(self):
+        return self.algorithm
+
+    def get_kid(self):
+        return self.kid
+
+# Retrieve sensitive data from environment variables
+storage_url = get_env_var(STORAGE_URL)
+keyvault_url = get_env_var(KEYVAULT_URL)
+
+# Retrieve service principal values from environment variables
+# The service principal can be created using Azure CLI's `az ad create-for-rbac` command.
+tenant_id = get_env_var(TENANT_ID)
+client_id = get_env_var(CLIENT_ID)  # aka appId in AzureCLI
+client_secret = get_env_var(CLIENT_SECRET)  # aka password in AzureCLI
+
+# construct a token credential for use by Storage and KeyVault clients.
+credential = ClientSecretCredential(tenant_id, client_id, client_secret)
+secret_client = SecretClient(keyvault_url, credential=credential)
+
+# the secret is url-safe base64 encoded bytes, content type 'application/octet-stream'
+secret = secret_client.get_secret('symmetric-key')
+key_bytes = base64.urlsafe_b64decode(secret.value)
+kvk = KeyVaultKey(key_id=secret.id, key_ops=['unwrapKey', 'wrapKey'], k=key_bytes, kty='oct')
+kek = KeyWrapper(kvk, credential)
+
+storage_client = BlobServiceClient(storage_url, credential=credential)
+container_name = make_resource_name('container')
+blob_name = make_resource_name('blob')
+
+container_client = storage_client.get_container_client(container_name)
+# if kek is used, error: "wrap_key" is missing or invalid
+# if crypto_client is used, error: "get_kid" is missing or invalid
```

Comment:
Vestigial comments?

Reply:
Ah, yes, I was trying to use the KeyVault primitives directly and neither worked for these reasons. I had to write the KeyWrapper class to conform to the Storage SDK's desired interface.